tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,511 @@
+}
+
+func (w *HeadReadWriter) chunkRef(seq, offset uint64) (chunkRef uint64) {
+	return (seq << 32) | offset
+}
+
+func (w *HeadReadWriter) shouldCutSegment(chunkLength int) bool {
+	return w.n == 0 || // First segment
+		// TODO: tune this boolean, cutting a segment for only 1 chunk would be inefficient.
+		(time.Since(w.curFileStartTime) > w.segmentTime && w.n > HeadSegmentHeaderSize) || // Time duration reached for the existing file.
+		w.n+int64(chunkLength+27) >= MaxSegmentSize
+}
+
+func (w *HeadReadWriter) seq() int {
+	return w.curFileSequence
+}
+
+func (w *HeadReadWriter) write(b []byte) error {
+	n, err := w.wbuf.Write(b)
+	w.n += int64(n)
+	return err
+}
+
+func (w *HeadReadWriter) cut() (err error) {
+	// Sync current tail to disk and close.
+	if err := w.finalizeCurFile(); err != nil {
+		return err
+	}
+
+	n, f, seq, err := cutSegmentFile(w.dirFile, headChunksFormatV1, false, 0)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			f.Close()
+		}
+	}()
+
+	w.n = int64(n)
+	oldSeq := w.curFileSequence
+	w.curFileSequence = seq
+
+	// Write current time in milliseconds for mint and 0 for maxt.
+	w.curFileStartTime = time.Now()
+	binary.BigEndian.PutUint64(w.buf[:], uint64(w.curFileStartTime.UnixNano()/1e6))
+	if _, err := f.Write(w.buf[:8]); err != nil {
+		return err
+	}
+	w.n += 8
+	binary.BigEndian.PutUint64(w.buf[:], 0)
+	if _, err := f.Write(w.buf[:8]); err != nil {
+		return err
+	}
+	w.n += 8
+
+	oldFile := w.curFile
+
+	w.curFile = f
+	if w.wbuf != nil {
+		w.wbuf.Reset(f)
+	} else {
+		w.wbuf = bufio.NewWriterSize(f, 1024)
+	}
+
+	if oldFile != nil {
+		// Open it again with the new size.
+		newTailFile, err := fileutil.OpenMmapFile(oldFile.Name())
+		if err != nil {
+			return err
+		}
+		w.bsMtx.Lock()
+		// Closing the last mmapped file.
+		if err := w.cs[oldSeq].Close(); err != nil {
+			w.bsMtx.Unlock()
+			return err
+		}
+		w.cs[oldSeq] = newTailFile
+		w.bs[oldSeq] = realByteSlice(newTailFile.Bytes())
+		w.bsMtx.Unlock()
+	}
+
+	mmapFile, err := fileutil.OpenMmapFileWithSize(f.Name(), int(MaxSegmentSize))
+	if err != nil {
+		return err
+	}
+	w.bsMtx.Lock()
+	w.cs[w.curFileSequence] = mmapFile
+	w.bs[w.curFileSequence] = realByteSlice(mmapFile.Bytes())
+	w.bsMtx.Unlock()
+
+	return nil
+}
+
+// Truncate deletes the segment files which are strictly below the mint.
+// mint should be in milliseconds.
+func (w *HeadReadWriter) Truncate(mint int64) error {
+	var removedFiles []int
+
+	w.bsMtx.RLock()
+	for seq, bs := range w.bs {
+		b := bs.Range(HeaderMaxtOffset, HeaderMaxtOffset+8)
+		maxt := binary.BigEndian.Uint64(b)
+		if maxt != 0 && int64(maxt) < mint {
+			removedFiles = append(removedFiles, seq)
+		}
+	}
+	w.bsMtx.RUnlock()
+
+	closers := make([]io.Closer, 0, len(removedFiles))
+	w.bsMtx.Lock()
+	for _, seq := range removedFiles {
+		closers = append(closers, w.cs[seq])
+		delete(w.bs, seq)
+		delete(w.cs, seq)
+		if err := os.Remove(segmentFile(w.dirFile.Name(), seq)); err != nil {
```

Comment:
What about the `w.size`? Should we substract the size of those?

Reply:
Good point. We should not only subtract here but also add when we cut a new segment file :) Thanks!